gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/SelectionAndFilterNode.java

```diff
@@ -0,0 +1,53 @@
+package org.hypertrace.gateway.service.entity.query;
+
+import java.util.List;
+import org.hypertrace.gateway.service.entity.query.visitor.Visitor;
+import org.hypertrace.gateway.service.v1.common.OrderByExpression;
+
+public class SelectionAndFilterNode implements QueryNode {
+
+  private final String source;
+  private final int limit;
+  private final int offset;
+  private final List<OrderByExpression> orderBys;
+
+  public SelectionAndFilterNode(String source, int limit, int offset, List<OrderByExpression> orderBys) {
+    this.source = source;
+    this.limit = limit;
+    this.offset = offset;
+    this.orderBys = orderBys;
```

Comment:
orderBys is not read anywhere.
I actually dont think this new SelectAndFilterNode is even necessary. Here is what i think is ideal:
- Merge all the getEntities, getAggregations, getEntitiesAndAggregations to one method
- Enhance the existing DataFetcherNode to add limit, offset and orderBys (but all are optional). Depending on whether the query is single source vs multiple we set these fields, like how we are conditionally creating a DataFetcherNode vs SelectionAndFilterNode

So in the QueryServiceEntityFetcher while constructing the query-service query we add whatever checks are needed for limit, offset, orderBy since they are optional and the ExecutionTree building logic is the one that decided whether to set them or not.
That way the ExecutionVisitor doesn't have to change at all.

I may be missing something here. But happy to discuss over a call if that helps

cc: @inespo @buchi-busireddy

Reply:
- I can remove the unused orderBys.
- I can merge getEntities, getAggregations, getEntitiesAndAggregations to one method but will need robust unit tests before doing that. I added some for QueryServiceEntityFetcher and will need to add more to cover the complexity. So not in this PR.
- I will not make DataFetcherNode or QueryServiceEntityFetcher more complex than it already is. Every time you add an if else, makes it harder to understand the code. We should strive to reduce the complexity in ExecutionVisitor by not adding more branches to one method. QueryServiceEntityFetcher should not have to determine what orderby, offset or limit to set. It's work should just be to take the EntitiesRequest as is and translate it to a QueryService request, execute it and parse the response to an EntityService response.
- I will rename DataFetcherNode to FilterNode and probably rename SelectAndFilterNode to SelectFilterAndOrderNode or SingleSourceNode or something.

Reply:
> QueryServiceEntityFetcher should not have to determine what orderby, offset or limit to set. It's work should just be to take the EntitiesRequest as is and translate it to a QueryService request
This is precisely why I think all the different methods in QueryServiceEntityFetcher should be merged to one and I strongly believe it should be possible. 
- orderBy being an empty list should and i believe is already gracefully handled
- offset i believe the default is 0 (so it should just work)
- limit is the only one which if not set in the EntitiesRequest, we need to set it to some upper limit(just as a safety)

Once you do that, i think both the DataFetcherNode and SelectAnFilterNode are almost the same and hence can be merged. I think having an extra SelectAnFilterNode makes the ExecutionVisitor even more complex to understand. 

Since, you are anyways going through the process of testing the change(not sure how, assuming manually), may be we should just do all the changes in one shot and test.

Reply:
I guess the prerequisite for merging the SelectAndFilterNode and DataFetcherNode is a single method for getEntities in QueryServiceEntityFetcher, which you plan to take up in a different PR once we have more tests.

So, we can probably defer that too, but I'm just nervous because the focus on this will be diverted once the perf issue is resolved. So, leaving it to your judgement.